File: sdks/typescript/src/onboarding/propose.ts
/**
 * Propose — turn a scan into UNCONFIRMED candidate outcome rules. The TS mirror of Python
 * `valuemaxx.onboarding.propose` (a golden parity test asserts the two produce the same
 * proposal). The signal class is SYSTEM-owned (a bare external write is only
 * `action_attempted`; an in-process status/mark/ORM write, a webhook, confirm an outcome) —
 * never user-set, so a function attempt can never masquerade as a confirmed outcome.
 */

import { redact } from "./redact.js";
import type { OnboardingRules, ScanSite } from "./types.js";

type MatchKind = "status_setter" | "mark_function" | "orm_write" | "external_write" | "webhook";
type BindingTier = "exact" | "deterministic" | "candidate" | "likely";
type SignalClass = "action_attempted" | "outcome_confirmed";

export interface RunIdInjection {
  readonly system: string;
  readonly targetField: string;
  readonly writeSite: string;
}

export interface OutcomeRuleCandidate {
  readonly name: string;
  readonly matchKind: MatchKind;
  readonly matchTarget: string;
  readonly when: string;
  readonly signal: SignalClass;
  readonly tier: BindingTier;
  readonly runIdInjection: RunIdInjection | null;
  readonly warnings: readonly string[];
}

export interface Proposal {
  readonly rules: readonly OutcomeRuleCandidate[];
  readonly entityIds: readonly string[];
  readonly sharedCostsPresent: boolean;
  readonly warnings: readonly string[];
}

const SITE_TO_MATCH: Record<string, MatchKind> = {
  status_setter: "status_setter",
  mark_function: "mark_function",
  orm_write: "orm_write",
  external_write: "external_write",
  webhook_handler: "webhook",
};

// Site kinds that bind synchronously in-process (exact).
const IN_PROCESS_KINDS = new Set(["status_setter", "mark_function", "orm_write"]);

// Match kinds whose signal is a CONFIRMED outcome (system-owned; mirrors the Python mapper).
const CONFIRMING = new Set<MatchKind>(["status_setter", "mark_function", "orm_write", "webhook"]);

// The run-id metadata field injected into an echoing system's outbound object.
const INJECTED_FIELD = "metadata.atm_run_id";

function signalFor(site: ScanSite): SignalClass {
  return CONFIRMING.has(SITE_TO_MATCH[site.kind]!) ? "outcome_confirmed" : "action_attempted";
}

function defaultWhen(site: ScanSite): string {
  if (site.kind === "status_setter") return "args.status != None";
  if (site.kind === "webhook_handler") return "event.type != None";
  return "True";
}

function ruleForSite(site: ScanSite, rules: OnboardingRules): OutcomeRuleCandidate {
  const matchKind = SITE_TO_MATCH[site.kind]!;
  const signal = signalFor(site);
  let tier: BindingTier;
  let injection: RunIdInjection | null = null;
  let warnings: string[] = [];

  if (IN_PROCESS_KINDS.has(site.kind)) {
    tier = "exact";
  } else if (site.kind === "webhook_handler") {
    tier = "deterministic";
  } else if (site.kind === "external_write" && site.echoesMetadata) {
    tier = "deterministic";
    injection = {
      system: redact(site.system ?? "unknown", rules),
      targetField: INJECTED_FIELD,
      writeSite: redact(site.symbol, rules),
    };
  } else {
    // non-echoing external write
    tier = "candidate";
    const system = redact(site.system ?? "unknown", rules);
    warnings = [
      `${system} does not echo injected metadata; deterministic T3 binding is ` +
        `unavailable. Falling back to entity-id matching (candidate/T4); a human ` +
        `must confirm before this is trusted.`,
    ];
  }

  return {
    name: redact(site.symbol, rules),
    matchKind,
    matchTarget: redact(`${site.file}:${site.symbol}`, rules),
    when: redact(defaultWhen(site), rules),
    signal,
    tier,
    runIdInjection: injection,
    warnings,
  };
}

/** Build a reviewable proposal of UNCONFIRMED candidate rules from a scan. */
export function buildProposal(
  scan: {
    outcomeSites: readonly ScanSite[];
    entityIds: readonly string[];
    warnings: readonly string[];
  },
  rules: OnboardingRules,
): Proposal {
  return {
    rules: scan.outcomeSites.map((s) => ruleForSite(s, rules)),
    entityIds: scan.entityIds.map((e) => redact(e, rules)),
    sharedCostsPresent: false,
    warnings: scan.warnings.map((w) => redact(w, rules)),
  };
}

File: packages/onboarding/src/valuemaxx/onboarding/propose.py
"""PROPOSE — turn a scan into UNCONFIRMED candidate outcome rules (design §7 step 2).

For each discovered outcome site, :func:`build_proposal` drafts an
:class:`~valuemaxx.onboarding.capabilities.OutcomeRuleCandidate`:

* in-process sites (status setter / mark_* / ORM write) → ``exact`` binding (the
  rule fires synchronously at the call site);
* echoing external writes (Stripe/HubSpot/Zendesk) → a declarative ``run_id``
  injection block + ``deterministic`` binding (the later webhook echoes it back);
* non-echoing external writes (Salesforce …) → no injection, ``candidate`` binding,
  and a warning naming the system (T4 entity-fallback territory);
* webhook handlers → ``deterministic`` binding (the inbound payload carries the id).

The signal class is always the system-mapped value (``signal_mapper.map_signal``,
the :class:`~valuemaxx.core.SignalClassMapper` seam) — never user-set, so a function
attempt can never masquerade as a confirmed outcome. Every rule is ``confirmed=False``
(human review is the only path to confirmation), and every captured string is
re-redacted defensively before it enters a proposal field.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

from valuemaxx.core import BindingTier, SignalClass
from valuemaxx.core.wire import INJECTED_RUN_ID_FIELD
from valuemaxx.onboarding.capabilities import (
    MatchKind,
    OutcomeRuleCandidate,
    Proposal,
    RunIdInjection,
    ScanSite,
)
from valuemaxx.onboarding.redact import redact

if TYPE_CHECKING:
    from valuemaxx.core import SignalClassMapper
    from valuemaxx.onboarding.capabilities import ScanResult

# Map a discovered site kind to the outcomes.yaml match kind.
_SITE_TO_MATCH: dict[str, MatchKind] = {
    "status_setter": "status_setter",
    "mark_function": "mark_function",
    "orm_write": "orm_write",
    "external_write": "external_write",
    "webhook_handler": "webhook",
}

# Site kinds that bind synchronously in-process (exact).
_IN_PROCESS_KINDS = frozenset({"status_setter", "mark_function", "orm_write"})

# The run-id metadata field injected into an echoing system's outbound object. Sourced
# from the cross-language wire contract so onboard's proposal and the runtime T3 injector
# (both SDKs) default to the identical path — they cannot drift.
_INJECTED_FIELD = INJECTED_RUN_ID_FIELD


def _signal_for(site: ScanSite, mapper: SignalClassMapper) -> SignalClass:
    """The system-owned signal class for a site (never user-set)."""
    match_kind = _SITE_TO_MATCH[site.kind]
    return SignalClass(mapper.map_signal(match_kind=match_kind, declared=""))


def _default_when(site: ScanSite) -> str:
    """A conservative default predicate for the site (redacted, human-editable)."""
    if site.kind == "status_setter":
        return "args.status != None"
    if site.kind == "webhook_handler":
        return "event.type != None"
    return "True"


def _rule_for_site(site: ScanSite, mapper: SignalClassMapper) -> OutcomeRuleCandidate:
    """Draft one unconfirmed candidate rule for a discovered site."""
    match_kind = _SITE_TO_MATCH[site.kind]
    signal = _signal_for(site, mapper)
    injection: RunIdInjection | None = None
    warnings: tuple[str, ...] = ()

    if site.kind in _IN_PROCESS_KINDS:
        tier = BindingTier.EXACT
    elif site.kind == "webhook_handler":
        tier = BindingTier.DETERMINISTIC
    elif site.kind == "external_write" and site.echoes_metadata:
        tier = BindingTier.DETERMINISTIC
        injection = RunIdInjection(
            system=redact(site.system or "unknown"),
            target_field=_INJECTED_FIELD,
            write_site=redact(site.symbol),
        )
    else:  # non-echoing external write
        tier = BindingTier.CANDIDATE
        system = redact(site.system or "unknown")
        warnings = (
            f"{system} does not echo injected metadata; deterministic T3 binding is "
            f"unavailable. Falling back to entity-id matching (candidate/T4); a human "
            f"must confirm before this is trusted.",
        )

    return OutcomeRuleCandidate(
        name=redact(site.symbol),
        match_kind=match_kind,
        match_target=redact(f"{site.file}:{site.symbol}"),
        when=redact(_default_when(site)),
        signal=signal,
        tier=tier,
        run_id_injection=injection,
        warnings=warnings,
    )


def build_proposal(
    scan: ScanResult,
    *,
    signal_mapper: SignalClassMapper,
    shared_costs_inputs: bool = False,
) -> Proposal:
    """Build a reviewable :class:`Proposal` of UNCONFIRMED candidate rules from a scan.

    ``shared_costs_inputs`` is True only when the operator has Tier-2/3 cost inputs
    (GPU seconds, monthly bills …); absent those, ``shared_costs_present`` stays
    False and no shared_costs.yaml is implied (M6 — never report a partial number as
    complete).
    """
    rules = tuple(_rule_for_site(site, signal_mapper) for site in scan.outcome_sites)
    warnings = tuple(redact(w) for w in scan.warnings)
    return Proposal(
        rules=rules,
        entity_ids=tuple(redact(e) for e in scan.entity_ids),
        shared_costs_present=shared_costs_inputs,
        warnings=warnings,
    )


__all__ = ["build_proposal"]
